### PR TITLE
GH-3026 Support chmod with FTP

### LINKED ISSUE #3026: can't chmod using ftp outbound Adapter
**Affects Version(s):** 5.2.0.M4

---
## Question

Hello guys, I'm having problems to use chmod with an outbound adapter,  how can this be solved?

I gave a look around the code class **FileTransferringMessageHandler,** seems to be hardcoded
image

my dsl

```java

  return flow -> flow
            .handle(Ftp.outboundGateway(FTPServers.PC_LOCAL.getFactory(), AbstractRemoteFileOutboundGateway.Command.MGET, "payload")
                .fileExistsMode(FileExistsMode.REPLACE)
                .filterFunction(ftpFile -> ftpFile.getName().contains("SCN") || ftpFile.getName().contains("REP") || ftpFile.getName().contains("EL"))
                .localDirectory(new File("/localTmp")))
            .log(Message::getPayload)
            .transform(GenericMessage.class, GenericMessage::getPayload)
            .split()
            .handle(Ftp.outboundAdapter(FTPServers.PC_LOCAL.getFactory(), FileExistsMode.IGNORE)
                .useTemporaryFileName(false)
                .chmod(666)
                .fileNameExpression("headers['" + FileHeaders.REMOTE_FILE + "']")
                .remoteDirectory("/dati/recpv"), e -> e.advice(afterFlow()));
    }
```

### PR DESCRIPTION
**Pull request by garyrussell**
Resolves https://github.com/spring-projects/spring-integration/issues/3026

**cherry-pick to 5.1.x**

### CODE REVIEW

File: spring-integration-ftp/src/main/java/org/springframework/integration/ftp/gateway/FtpOutboundGateway.java [hunk 1]
```diff
@@ -299,4 +301,22 @@ else if (e instanceof RuntimeException) {
...
+
+	@Override
+	protected void doChmod(RemoteFileOperations<FTPFile> remoteFileOperations, final String path, final int chmod) {
+		remoteFileOperations.executeWithClient((ClientCallbackWithoutResult<FTPClient>) client -> {
+			String chModCommand = "chmod " + Integer.toOctalString(chmod) + " " + path;
+			try {
+				client.sendSiteCommand(chModCommand);
+			}
+			catch (IOException e) {
+				throw new UncheckedIOException("Failed to execute chmod", e);
```
Comment (artembilan) [inline spring-integration-ftp/src/main/java/org/springframework/integration/ftp/gateway/FtpOutboundGateway.java]: Don't we need to add a `chmod` and `path` into the exception message as well?

File: spring-integration-ftp/src/test/java/org/springframework/integration/ftp/outbound/FtpServerOutboundTests.java [hunk 2]
```diff
@@ -399,7 +402,13 @@ public void testRawGETWithTemplate() throws Exception {
 	}
 
 	@Test
-	public void testInt3088MPutNotRecursive() {
+	// TODO: Test on Windows - if fails, @DisableIf or stub method on the spy
```
Comment (artembilan) [inline spring-integration-ftp/src/test/java/org/springframework/integration/ftp/outbound/FtpServerOutboundTests.java]: So, this works for me on Windows.

**Review by artembilan — CHANGES_REQUESTED** [review]
Gary, LGTM from big height.
I can polish those simple concerns on merge.
Just let me know!

**Review by artembilan — CHANGES_REQUESTED** [review]
Would you mind to fix conflicts before I merge this:
```

Conflicting files
src/reference/asciidoc/whats-new.adoc
```
?

Thanks

**Comment by garyrussell** [comment]
rebased and resolved

**Comment by artembilan** [comment]
Cherry-picked to `5.1.x` as https://github.com/spring-projects/spring-integration/commit/b04a8576fab9031ea4cbe3aa3d80651b01a41c82 after resolving conflicts.

**Pull request merged**